{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d50015a1-f2aa-4d53-a462-0bec5b15c34c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/srv/conda/lib/python3.11/site-packages/Bio/Application/__init__.py:39: BiopythonDeprecationWarning: The Bio.Application modules and modules relying on it have been deprecated.\n",
      "\n",
      "Due to the on going maintenance burden of keeping command line application\n",
      "wrappers up to date, we have decided to deprecate and eventually remove these\n",
      "modules.\n",
      "\n",
      "We instead now recommend building your command line and invoking it directly\n",
      "with the subprocess module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# GENERATE alignment files for HA\n",
    "\n",
    "#Extract the HA sequences\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def extract_HA_sequences(fasta_files, gff_files, output_file):\n",
    "    ha_sequences = []\n",
    "    names = [\"flu_2018\", \"flu_2021\", \"flu_2022\"]\n",
    "    \n",
    "    for fasta_file, gff_file, name in zip(fasta_files, gff_files, names):\n",
    "        genome_dict = {record.id: record.seq for record in SeqIO.parse(fasta_file, \"fasta\")}\n",
    "        with open(gff_file) as gff:\n",
    "            for line in gff:\n",
    "                if not line.startswith(\"#\"):\n",
    "                    fields = line.strip().split(\"\\t\")\n",
    "                    if len(fields) > 2 and fields[2] == \"gene\" and \"HA\" in fields[8]:\n",
    "                        seq_id = fields[0]\n",
    "                        start, end = int(fields[3]) - 1, int(fields[4])\n",
    "                        ha_seq = genome_dict[seq_id][start:end]\n",
    "                        ha_sequences.append((name, ha_seq))\n",
    "                        break\n",
    "    \n",
    "    with open(output_file, \"w\") as out_fasta:\n",
    "        for name, seq in ha_sequences:\n",
    "            out_fasta.write(f\">{name}\\n{seq}\\n\")\n",
    "\n",
    "# Input files\n",
    "fasta_files = [\"flu_2018.fa\", \"flu_2021.fa\", \"flu_2022.fa\"]\n",
    "gff_files = [\"flu_2018.gff\", \"flu_2021.gff\", \"flu_2022.gff\"]\n",
    "output_fasta = \"HA_sequences.fasta\"\n",
    "\n",
    "# Extract HA sequences\n",
    "extract_HA_sequences(fasta_files, gff_files, output_fasta)\n",
    "\n",
    "\n",
    "\n",
    "#____________________________________________________________________________________________________________________________\n",
    "#Get Mafft to align the sequences\n",
    "\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "import os\n",
    "\n",
    "def align_sequences(input_file, output_file=\"aligned_sequences.fasta\"):\n",
    "    mafft_cline = MafftCommandline(input=input_file)\n",
    "    stdout, stderr = mafft_cline()\n",
    "    with open(output_file, \"w\") as aligned:\n",
    "        aligned.write(stdout)\n",
    "    return output_file\n",
    "\n",
    "aligned_file = align_sequences(\"HA_sequences.fasta\", \"HA_aligned_sequences.fasta\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "569be6d0-3b60-4090-8c75-c6f71bacd8c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENERATE alignment files for NA\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def extract_NA_sequences(fasta_files, gff_files, output_file):\n",
    "    na_sequences = []\n",
    "    names = [\"flu_2018\", \"flu_2021\", \"flu_2022\"]\n",
    "    \n",
    "    for fasta_file, gff_file, name in zip(fasta_files, gff_files, names):\n",
    "        genome_dict = {record.id: record.seq for record in SeqIO.parse(fasta_file, \"fasta\")}\n",
    "        with open(gff_file) as gff:\n",
    "            for line in gff:\n",
    "                if not line.startswith(\"#\"):\n",
    "                    fields = line.strip().split(\"\\t\")\n",
    "                    if len(fields) > 2 and fields[2] == \"gene\" and \"NA\" in fields[8]:\n",
    "                        seq_id = fields[0]\n",
    "                        start, end = int(fields[3]) - 1, int(fields[4])\n",
    "                        na_seq = genome_dict[seq_id][start:end]\n",
    "                        na_sequences.append((name, na_seq))\n",
    "                        break\n",
    "    \n",
    "    with open(output_file, \"w\") as out_fasta:\n",
    "        for name, seq in na_sequences:\n",
    "            out_fasta.write(f\">{name}\\n{seq}\\n\")\n",
    "\n",
    "# Input files\n",
    "fasta_files = [\"flu_2018.fa\", \"flu_2021.fa\", \"flu_2022.fa\"]\n",
    "gff_files = [\"flu_2018.gff\", \"flu_2021.gff\", \"flu_2022.gff\"]\n",
    "output_fasta = \"NA_sequences.fasta\"\n",
    "\n",
    "# Extract NA sequences\n",
    "extract_NA_sequences(fasta_files, gff_files, output_fasta)\n",
    "\n",
    "\n",
    "#____________________________________________________________________________________________________________________________\n",
    "#Get Mafft to align the sequences\n",
    "\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "import os\n",
    "\n",
    "def align_sequences(input_file, output_file=\"aligned_sequences.fasta\"):\n",
    "    mafft_cline = MafftCommandline(input=input_file)\n",
    "    stdout, stderr = mafft_cline()\n",
    "    with open(output_file, \"w\") as aligned:\n",
    "        aligned.write(stdout)\n",
    "    return output_file\n",
    "\n",
    "aligned_file = align_sequences(\"NA_sequences.fasta\", \"NA_aligned_sequences.fasta\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cb069b47-12bb-4ec5-90fc-bcb644e0f4b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENERATE alignment files for PB1 \n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def extract_pb1_sequences(fasta_files, gff_files, output_file):\n",
    "    pb1_sequences = []\n",
    "    names = [\"flu_2018\", \"flu_2021\", \"flu_2022\"]\n",
    "    \n",
    "    for fasta_file, gff_file, name in zip(fasta_files, gff_files, names):\n",
    "        genome_dict = {record.id: record.seq for record in SeqIO.parse(fasta_file, \"fasta\")}\n",
    "        with open(gff_file) as gff:\n",
    "            for line in gff:\n",
    "                if not line.startswith(\"#\"):\n",
    "                    fields = line.strip().split(\"\\t\")\n",
    "                    if len(fields) > 2 and fields[2] == \"gene\" and \"PB1\" in fields[8]:\n",
    "                        seq_id = fields[0]\n",
    "                        start, end = int(fields[3]) - 1, int(fields[4])\n",
    "                        pb1_seq = genome_dict[seq_id][start:end]\n",
    "                        pb1_sequences.append((name, pb1_seq))\n",
    "                        break\n",
    "    \n",
    "    with open(output_file, \"w\") as out_fasta:\n",
    "        for name, seq in pb1_sequences:\n",
    "            out_fasta.write(f\">{name}\\n{seq}\\n\")\n",
    "\n",
    "# Input files\n",
    "fasta_files = [\"flu_2018.fa\", \"flu_2021.fa\", \"flu_2022.fa\"]\n",
    "gff_files = [\"flu_2018.gff\", \"flu_2021.gff\", \"flu_2022.gff\"]\n",
    "output_fasta = \"pb1_sequences.fasta\"\n",
    "\n",
    "# Extract PB1 sequences\n",
    "extract_pb1_sequences(fasta_files, gff_files, output_fasta)\n",
    "\n",
    "\n",
    "#____________________________________________________________________________________________________________________________\n",
    "#Get Mafft to align the sequences\n",
    "\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "import os\n",
    "\n",
    "def align_sequences(input_file, output_file=\"aligned_sequences.fasta\"):\n",
    "    mafft_cline = MafftCommandline(input=input_file)\n",
    "    stdout, stderr = mafft_cline()\n",
    "    with open(output_file, \"w\") as aligned:\n",
    "        aligned.write(stdout)\n",
    "    return output_file\n",
    "\n",
    "aligned_file = align_sequences(\"pb1_sequences.fasta\", \"pb1_aligned_sequences.fasta\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7adac4a9-1de4-438d-9218-792aef066720",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENERATE alignment files for PB2\n",
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def extract_pb2_sequences(fasta_files, gff_files, output_file):\n",
    "    pb2_sequences = []\n",
    "    names = [\"flu_2018\", \"flu_2021\", \"flu_2022\"]\n",
    "    \n",
    "    for fasta_file, gff_file, name in zip(fasta_files, gff_files, names):\n",
    "        genome_dict = {record.id: record.seq for record in SeqIO.parse(fasta_file, \"fasta\")}\n",
    "        with open(gff_file) as gff:\n",
    "            for line in gff:\n",
    "                if not line.startswith(\"#\"):\n",
    "                    fields = line.strip().split(\"\\t\")\n",
    "                    if len(fields) > 2 and fields[2] == \"gene\" and \"PB2\" in fields[8]:\n",
    "                        seq_id = fields[0]\n",
    "                        start, end = int(fields[3]) - 1, int(fields[4])\n",
    "                        pb2_seq = genome_dict[seq_id][start:end]\n",
    "                        pb2_sequences.append((name, pb2_seq))\n",
    "                        break\n",
    "    \n",
    "    with open(output_file, \"w\") as out_fasta:\n",
    "        for name, seq in pb2_sequences:\n",
    "            out_fasta.write(f\">{name}\\n{seq}\\n\")\n",
    "\n",
    "# Input files\n",
    "fasta_files = [\"flu_2018.fa\", \"flu_2021.fa\", \"flu_2022.fa\"]\n",
    "gff_files = [\"flu_2018.gff\", \"flu_2021.gff\", \"flu_2022.gff\"]\n",
    "output_fasta = \"pb2_sequences.fasta\"\n",
    "\n",
    "# Extract PB2 sequences\n",
    "extract_pb2_sequences(fasta_files, gff_files, output_fasta)\n",
    "\n",
    "\n",
    "#____________________________________________________________________________________________________________________________\n",
    "#Get Mafft to align the sequences\n",
    "\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "import os\n",
    "\n",
    "def align_sequences(input_file, output_file=\"aligned_sequences.fasta\"):\n",
    "    mafft_cline = MafftCommandline(input=input_file)\n",
    "    stdout, stderr = mafft_cline()\n",
    "    with open(output_file, \"w\") as aligned:\n",
    "        aligned.write(stdout)\n",
    "    return output_file\n",
    "\n",
    "aligned_file = align_sequences(\"pb2_sequences.fasta\", \"pb2_aligned_sequences.fasta\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5832dfdf-51ab-4485-aa31-1d3da8026096",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
